aten/src/ATen/core/dispatch/Dispatcher.cpp

```diff
@@ -112,14 +81,34 @@ RegistrationHandleRAII Dispatcher::registerDef(FunctionSchema schema) {
 
   OperatorName op_name = schema.operator_name();
 
-  auto op = findOrRegisterSchema_(std::move(schema));
+  const auto found = findSchema(op_name);
+  OperatorHandle op = [&] {
+    if (found == c10::nullopt) {
+      // Initial def
+      operators_.emplace_back(std::move(schema));
+      OperatorHandle op(--operators_.end());
+      operatorLookupTable_.write([&] (ska::flat_hash_map<OperatorName, OperatorHandle>& operatorLookupTable) {
+        operatorLookupTable.emplace(op_name, op);
+      });
```

Comment:
Yes this should work.

Reply:
New code nicer!

```
  OperatorName op_name = schema.operator_name();
  auto op = findOrRegisterName_(op_name);
        
  if (op.operatorIterator_->def_count == 0) {
    // NB: registerSchema is not idempotent! Only do it once!
    op.operatorIterator_->op.registerSchema(std::move(schema));
    listeners_->callOnOperatorRegistered(op);
  } else {
    TORCH_CHECK(op.schema() == schema, "Tried to register multiple operators with the same name and the same overload name but different schemas: ", schema, " vs ", op.schema());
    if (schema.isDefaultAliasAnalysisKind()) {
      // If the *new* schema is the default alias analysis kind, for BC, we
      // will accept it.  If we don't accept it, most extensions that override
      // existing operators will stop working (as they generally did not
      // specify alias information).  Remove this BC smoothing ASAP, because
      // if the two incompatible registrations live in the same compilation
      // unit, the order their static initializers run is unspecified, which
      // means that you may nondeterministically fail the subsequent test.
    } else if (op.schema().isDefaultAliasAnalysisKind()) {
      // If you POST-FACTO specify a non-default alias analysis kind after
      // we already have a schema for a function, complain loudly about it
      // (because this new implementation doesn't support merging in this
      // way).
      TORCH_CHECK(op.schema().aliasAnalysis() == schema.aliasAnalysis(),
        "Tried to define the schema for ", toString(op_name),
        " multiple times without providing an explicit alias analysis kind at each registration site.  "
        "This was previously permitted, but is now not allowed.  You should either explicitly specify the "
        "correct alias analysis kind at each site [",
        toString(op.schema().isDefaultAliasAnalysisKind() ? schema.aliasAnalysis() : op.schema().aliasAnalysis()),  
        "], or use the new Module::impl() API, which permits you to omit the schema entirely when "
        "specifying further implementations of an operator");
    } else {
      TORCH_CHECK(op.schema().aliasAnalysis() == schema.aliasAnalysis(),
        "Tried to define the schema for ", toString(op_name), " with different alias analysis kinds: ",
        toString(op.schema().aliasAnalysis()), " vs ", toString(schema.aliasAnalysis()));
    }   
  }     
    
  // NB: do not increment the counts until AFTER error checking
  ++op.operatorIterator_->def_count;
  ++op.operatorIterator_->def_and_impl_count;
```

Reply:
😍😍😍